File: js/apl-wasm/src/LiveMap.ts
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 */

 /**
  * A Live object is a single APL Object that changes over time.  LiveMaps are
  * created and modified by ViewHosts.
  *
  * Inside of the APL document the LiveMap may be used normally in data-binding
  * contexts.  For example:
  *
  *     {
  *       "type": "Text",
  *       "text": "The live object is currently '${MyLiveMap}'"
  *     }
  *
  *  The same LiveMap may be used by multiple Context objects. This allows
  *  an application to create a single LiveMap to track a source of changing
  *  data and display it in multiple view hosts simultaneously.
  *
  *  Changing the key-value pairs in a live map will update all data-bound
  *  values currently in the component hierarchy that depend on those values.
  */
export class LiveMap {

    /**
     * Create a LiveMap with an initial Object
     * @param map The initial Map.
     * @return The LiveMap
     */
    public static create(map? : any) : LiveMap {
        return new LiveMap(map);
    }
    public liveMap : APL.LiveMap;

    private constructor(map? : any) {
        this.liveMap = Module.LiveMap.create(map);
    }

    /**
     * Check to see if there are no elements in the map
     * @return True if the map is empty
     */
    public empty() : boolean {
        return this.liveMap.empty();
    }

    /**
     * Clear all elements from the map
     */
    public clear() : void {
        this.liveMap.clear();
    }

    /**
     * Retrieve a value from the map.  If the key does not exist, a NULL object
     * will be returned.
     * @param key The key of the object to return.
     * @return The internal object.
     */
    public get(key : string) : any {
        return this.liveMap.get(key);
    }

    /**
     * Check to see if a key exists in the map.
     * @param key The key to search for.
     * @return True if the key exists.
     */
    public has(key : string) : boolean {
        return this.liveMap.has(key);
    }

    /**
     * Set a key-value pair in the map.
     * @param key The key to insert
     * @param value The value to store.
     */
    public set(key : string, value : string) : void {
        this.liveMap.set(key, value);
    }

    /**
     * Set a collection of values from a different map
     * @map The object map to copy values from.
     */
    public update(map : any) : void {
        this.liveMap.update(map);
    }

    /**
     * Replace the LiveMap with a new map.
     * @param map The new map to set.
     */
    public replace(map : any) : void {
        this.liveMap.replace(map);
    }

    /**
     * Remove a key from the map
     * @param key The key to remove
     * @return True if the key was found and removed.
     */
    public remove(key : string) : boolean {
        return this.liveMap.remove(key);
    }
}

File: wasm/src/livemap.cpp
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 */

#include "wasm/livemap.h"
#include "wasm/embindutils.h"

namespace apl {
namespace wasm {

namespace internal {

apl::LiveMapPtr 
LiveMapMethods::create(emscripten::val map) {
    // We need to move it afterwards, so
    auto initMap = emscripten::getObjectMapFromVal(map);
    if (initMap) {    
        return apl::LiveMap::create(std::move(*initMap));
    }

    return apl::LiveMap::create();
}

bool 
LiveMapMethods::empty(const apl::LiveMapPtr& liveMapPtr) {
    return liveMapPtr->empty();
}

void 
LiveMapMethods::clear(const apl::LiveMapPtr& liveMapPtr) {
    liveMapPtr->clear();
}

emscripten::val 
LiveMapMethods::get(const apl::LiveMapPtr& liveMapPtr, const std::string& key) {
    return emscripten::getValFromObject(liveMapPtr->get(key), nullptr);
}

bool 
LiveMapMethods::has(const apl::LiveMapPtr& liveMapPtr, const std::string& key) {
    return liveMapPtr->has(key);
}

void 
LiveMapMethods::set(const apl::LiveMapPtr& liveMapPtr, const std::string& key, emscripten::val value) {
    liveMapPtr->set(key, emscripten::getObjectFromVal(value));
}

void 
LiveMapMethods::update(const apl::LiveMapPtr& liveMapPtr, emscripten::val map) {
    auto transformed = emscripten::getObjectMapFromVal(map);
    if (transformed)
        liveMapPtr->update(*transformed);
}

void 
LiveMapMethods::replace(const apl::LiveMapPtr& liveMapPtr, emscripten::val map) {
    auto transformed = emscripten::getObjectMapFromVal(map);
    if (transformed)
        liveMapPtr->replace(std::move(*transformed));
}

bool 
LiveMapMethods::remove(const apl::LiveMapPtr& liveMapPtr, const std::string& key) {
    return liveMapPtr->remove(key);
}

} // namespace internal

EMSCRIPTEN_BINDINGS(apl_wasm_livemap) {
    emscripten::class_<apl::LiveMap>("LiveMap")
        .smart_ptr<apl::LiveMapPtr>("LiveMapPtr")
        .class_function("create", &internal::LiveMapMethods::create)
        .function("empty", &internal::LiveMapMethods::empty)
        .function("clear", &internal::LiveMapMethods::clear)
        .function("get", &internal::LiveMapMethods::get)
        .function("has", &internal::LiveMapMethods::has)
        .function("set", &internal::LiveMapMethods::set)
        .function("update", &internal::LiveMapMethods::update)
        .function("replace", &internal::LiveMapMethods::replace)
        .function("remove", &internal::LiveMapMethods::remove);
}

} // namespace wasm
} // namespace apl
